include/kernel/thread.h

```diff
@@ -145,7 +145,10 @@ static inline void commitTime(void)
         }
         NODE_STATE(ksCurSC)->scConsumed += NODE_STATE(ksConsumed);
     }
-    if (CONFIG_NUM_DOMAINS > 1) {
+    /* If we have switched domains, ksDomainTime has already been reset and is accounted for the
+     * old domain. If we have not switched domains, we still need to account for the time that
+     * has passed. */
+    if (CONFIG_NUM_DOMAINS > 1 && !switchedDomain) {
         assert(ksDomainTime > NODE_STATE(ksConsumed));
         if (NODE_STATE(ksConsumed) < ksDomainTime) {
             ksDomainTime -= NODE_STATE(ksConsumed);
```

Comment:
The charge to domain time still occurs after the point where a domain may get switched here so we can still have the issue where a thread is scheduled for some period in a domain when there is no time left in that domain. Although at least now the timeout should be set such that we immediately return to the kernel to switch domains after.

Reply:
According to the `assert`, not really -- the consumed time at this point should be smaller than the remaining domain time. They might be very close together, and in that case you would get the out-of-kernel-into-kernel behaviour you describe (and if the assert is wrong, you would get the same), but I think that is fine.

The assert should be correct, because we set the next timer interrupt always to the minimum of remaining domain time and thread budget (- kWCET factor).

Reply:
(this is if the domain is not switched -- if the domain *is* switched, we don't explicitly charge the domain time any more, because it has already been discarded. Since the timer IRQ must occur at the minimum of domain switch and thread budget, the thread can't have overrun the domain time, so that part is fine. It would be possible for that same thread (in the old domain) to now have almost 0 time left when the domain gets scheduled again, but I think that is what should happen).